<launch>
    <!-- 파라미터 설정 -->
    <arg name="algorithm" default="ppo" doc="RL algorithm: ppo or sac"/>
    <arg name="config_file" default="$(find pointcloud_to_grid)/config/rl_params.yaml"/>
    <arg name="model_dir" default="$(find pointcloud_to_grid)/models"/>
    <arg name="train_mode" default="true" doc="true for training, false for inference"/>
    <arg name="carla_host" default="localhost"/>
    <arg name="carla_port" default="2000"/>
    
    <!-- CARLA 설정 -->
    <param name="carla_host" value="$(arg carla_host)"/>
    <param name="carla_port" value="$(arg carla_port)"/>
    
    <!-- 훈련 모드 -->
    <group if="$(arg train_mode)">
        <node name="rl_trainer" pkg="pointcloud_to_grid" type="train_sigmoid.py" output="screen">
            <param name="algorithm" value="$(arg algorithm)"/>
            <param name="config_file" value="$(arg config_file)"/>
            <param name="model_dir" value="$(arg model_dir)"/>
        </node>
    </group>
    
    <!-- 추론 모드 -->
    <group unless="$(arg train_mode)">
        <!-- 기존 pointcloud_to_grid 노드 -->
        <include file="$(find pointcloud_to_grid)/launch/pointcloud_to_grid.launch"/>
        
        <!-- RL inference 서버 -->
        <node name="rl_sigmoid_server" pkg="pointcloud_to_grid" type="rl_inference_server.py" output="screen">
            <param name="model_path" value="$(arg model_dir)/trained_sigmoid_$(arg algorithm).zip"/>
            <param name="env_path" value="$(arg model_dir)/env_normalize_$(arg algorithm).pkl"/>
            <param name="algorithm" value="$(arg algorithm)"/>
        </node>
        
        <!-- RL 최적화 노드 (C++) -->
        <node name="rl_sigmoid_optimizer" pkg="pointcloud_to_grid" type="rl_sigmoid_optimizer" output="screen">
            <param name="use_rl_params" value="true"/>
            <param name="fallback_to_grid_search" value="true"/>
        </node>
        
        <!-- 시각화 -->
        <node name="rviz" pkg="rviz" type="rviz" args="-d $(find pointcloud_to_grid)/rviz/rl_sigmoid.rviz" if="$(eval arg('train_mode') == false)"/>
    </group>
    
    <!-- 공통 파라미터 -->
    <rosparam file="$(arg config_file)" command="load"/>
    
    <!-- 로그 디렉토리 설정 -->
    <param name="log_directory" value="$(find pointcloud_to_grid)/logs"/>
    
    <!-- TensorBoard 로그 -->
    <param name="tensorboard_log_dir" value="$(find pointcloud_to_grid)/tensorboard_logs"/>
</launch>
